File: 序号/120.三角形最小路径和.kt

fun minimumTotal(triangle: List<MutableList<Int>>): Int {
    for (i in 1 until triangle.size){
        // 三角形和方形不一样，路线只能索引相同，
        for (j in triangle[i].indices){
            triangle[i][j] += minOf(
                triangle[i-1].getOrNull(j-1)?:Int.MAX_VALUE,
                triangle[i-1].getOrNull(j)?:Int.MAX_VALUE
            )
        }
    }
    return triangle.last().minOrNull()!!
}

File: 序号/120.三角形最小路径和.cpp
class Solution {
 public:
  int minimumTotal(vector<vector<int>> &triangle) {
    // 每个点的路径由上一层两个点过来，取小的那个
    if(triangle.size()==1) return triangle[0][0];
    int ans = INT_MAX;
    for (int i = 1; i < triangle.size(); i++) {
      for (int j = 0; j < i + 1; j++) {
        if (j == 0) {
          triangle[i][j] += triangle[i - 1][j];
        } else if (j == i) {
          triangle[i][j] += triangle[i - 1][j - 1];
        } else {
          triangle[i][j] =
              min(triangle[i - 1][j], triangle[i - 1][j - 1]) + triangle[i][j];
        }
        if (i == triangle.size() - 1) {
          ans = min(ans, triangle[i][j]);
        }
      }
    }
    return ans;
  }
};